### PR TITLE
Implements RFD-0022 - OpenSSH-compatible Agent Forwarding

### PR DESCRIPTION
**Pull request by tcsc**
Prior to this change, `tsh` will only ever forward the internal key
agent managed by `tsh` to a remote machine.

This change allows a user to specify that `tsh` should forward either
the `tsh`-internal keystore, or the system key agent at `$SSH_AUTH_SOCK`.

This change also brings the `-A` command-line option into line with
OpenSSH.

For more info refer to RFD-0022.

See-Also: #1571

### CODE REVIEW

**Comment by russjones** [comment]
@nklaassen @fspmarshall Can you review?

**Comment by nklaassen** [comment]
Does the command line option need to be updated, or is this coming in another PR? Seems like it is still a bool, and probably needs an update to the help string as well https://github.com/gravitational/teleport/blob/trent/key-agent/tool/tsh/tsh.go#L327

I'm not sure how difficult this could be, but would it be feasible to test that the correct key agent is actually forwarded when this option is set?

I missed the RFD review, but to me the options "yes" and "local" seem a bit unnecessarily confusing. Maybe something like "system" and "tsh", with "yes" as an alias to "system" could be more clear? Just spitballing here, not sure if this is reasonable

**Comment by tcsc** [comment]
@nklaassen - the command line semantics are expected to change. From the RFD:

> The effect of this change is that the `tsh ssh` option `-A` will automatically acquire
> semantics in line with the OpenSSH client, while the backwards compatible behaviour
> can be activated with something like:
>
> ```bash
> $ tsh ssh -o "ForwardAgent local" root@example.com
> ```

We could potentially add something like `-A` means `ForwardAgent yes` and `-A=local` means `ForwardAgent local`, but given there is already a way to activate legacy behaviour I didn't think it worthwhile. Will look at the help string, though.

Re: the option value names, we're kinds stuck with "yes" and "no" as they are compatible with the OpenSSH ssh client options. Aliasing "yes" to "system" might be a useful affordance, but I'd be tempted to get feedback from the wild. I can take or leave the name "local", so if we want to change that I have no objections.

**Comment by nklaassen** [comment]
@tcsc ahh I was just confused on that point, I thought that `tsh ssh -A local` was meant to work, realizing now that wouldn't line up with OpenSSH

As for the option value names, I'm just finding "yes" vs "local" unintuitive - both options could map to either one in my mind.

**Review by nklaassen — APPROVED** [review]
Code changes look good to me

File: lib/client/api.go [hunk 1]
```diff
@@ -90,6 +90,16 @@ func ValidateAgentKeyOption(supplied string) error {
...
 }
 
+// AgentForwardingMode  describes how the user key agent will be forwarded
+// to a remote machine, if at all.
+type AgentForwardingMode int
+
+const (
+	ForwardAgentNo    AgentForwardingMode = iota
+	ForwardAgentYes   AgentForwardingMode = iota
+	ForwardAgentLocal AgentForwardingMode = iota
```
Comment (awly) [inline lib/client/api.go:100]:
```suggestion
	ForwardAgentNo    AgentForwardingMode = iota
	ForwardAgentYes
	ForwardAgentLocal
```

File: tool/tsh/options.go [hunk 2]
```diff
@@ -114,6 +125,25 @@ var AllOptions = map[string]map[string]bool{
...
+
+	case forwardAgentTextYes:
+		return client.ForwardAgentYes
+
+	case forwardAgentTextLocal:
+		return client.ForwardAgentLocal
+
+	default:
+		log.Errorf(
+			"Invalid agent forwarding mode \"%s\". Defaulting to \"%s\"",
```
Comment (awly) [inline tool/tsh/options.go:141]:
```suggestion
			"Invalid agent forwarding mode %q. Defaulting to %q",
```

**Review by awly — APPROVED** [review]
Just to confirm: you tested this by logging into a remote server and running `ssh-add -L`?

**Comment by tcsc** [comment]
> Just to confirm: you tested this by logging into a remote server and running `ssh-add -L`?

```bash
$ ./build/tsh ssh rowlf@auth-node
rowlf@ip-172-31-5-54:~$ ssh-add -l
Could not open a connection to your authentication agent.
rowlf@ip-172-31-5-54:~$ exit

$ ./build/tsh ssh -A rowlf@auth-node
rowlf@ip-172-31-5-54:~$ ssh-add -l
256 SHA256:dR1daFKmKLEDK055QNGbFi6zur2SSl+NOMnxzP4uBwQ trent@goteleport.com (ED25519)
2048 SHA256:5FPfhJMZH7x1yfzK9G3IZFqGkopy600C1AOR9sDH0A8 teleport:kermit (RSA-CERT)
2048 SHA256:5FPfhJMZH7x1yfzK9G3IZFqGkopy600C1AOR9sDH0A8 teleport:kermit (RSA)
2048 SHA256:AEML8fwmFdUazCzj+o1wEKBGIEaR33ei1zG4h5eZFU4 teleport:beaker (RSA-CERT)
2048 SHA256:AEML8fwmFdUazCzj+o1wEKBGIEaR33ei1zG4h5eZFU4 teleport:beaker (RSA)
rowlf@ip-172-31-5-54:~$ exit

$ ./build/tsh ssh -o "ForwardAgent local" rowlf@auth-node
rowlf@ip-172-31-5-54:~$ ssh-add -l
2048 SHA256:AEML8fwmFdUazCzj+o1wEKBGIEaR33ei1zG4h5eZFU4 teleport:beaker (RSA-CERT)
2048 SHA256:AEML8fwmFdUazCzj+o1wEKBGIEaR33ei1zG4h5eZFU4 teleport:beaker (RSA)
rowlf@ip-172-31-5-54:~$ exit
```

File: tool/tsh/options.go (hunk 2 above)
Comment (tcsc) [reply tool/tsh/options.go]: Done

File: lib/client/api.go (hunk 1 above)
Comment (tcsc) [reply lib/client/api.go]: Done

**Review by russjones — APPROVED** [review]
Bot.

File: lib/client/session.go [hunk 3]
```diff
@@ -200,6 +203,22 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
 	return sess, nil
 }
 
+// selectKeyAgent picks the appropriate key agent for forwarding to the
+// server, if any.
+func selectKeyAgent(tc *TeleportClient) agent.Agent {
+	switch tc.ForwardAgent {
+	case ForwardAgentYes:
+		log.Debugf("Selecting System Key Agent")
```
Comment (klizhentas) [inline lib/client/session.go:211]: Nit: (derived from https://github.com/golang/go/wiki/CodeReviewComments#error-strings). Let's make our log lines proper english sentences, so:

`Selecting system key agent.`

File: tool/tsh/options.go [hunk 4]
```diff
@@ -114,6 +125,25 @@ var AllOptions = map[string]map[string]bool{
...
+
+	case forwardAgentTextYes:
+		return client.ForwardAgentYes
+
+	case forwardAgentTextLocal:
+		return client.ForwardAgentLocal
+
+	default:
+		log.Errorf(
+			"Invalid agent forwarding mode %q. Defaulting to %q",
```
Comment (klizhentas) [inline tool/tsh/options.go:141]: Same here.

**Review by klizhentas — APPROVED** [review]
Couple of nitpicks, but otherwise LGTM.  You created a beautifully written [RFD](https://github.com/gravitational/teleport/blob/master/rfd/0022-ssh-agent-forwarding.md) btw, especially the diagram, I enjoyed reading it.

File: tool/tsh/options.go (hunk 4 above)
Comment (tcsc) [reply tool/tsh/options.go:141]: Saw this after hitting Merge, will fix in new PR

File: lib/client/session.go (hunk 3 above)
Comment (tcsc) [reply lib/client/session.go:211]: Saw this after hitting Merge, will fix in new PR

**Pull request merged**